enterprise/server/cmd/ci_runner/main.go

```diff
@@ -753,79 +766,117 @@ func setupGitRepo(ctx context.Context) error {
+	// Delete the pushed branch ref if it already exists.
+	if err := git(ctx, &ws.log, "branch", "--delete", "--force", *pushedBranch); err != nil && !isBranchNotFound(err) {
+		return err
 	}
+	// Checkout the branch locally.
+	ref := fmt.Sprintf("%s/%s", gitRemoteName(*pushedRepoURL), *pushedBranch)
+	if err := git(ctx, &ws.log, "checkout", "--force", "--track", ref); err != nil {
+		return err
+	}
+	// Merge the target branch (if different from the pushed branch) so that the
+	// workflow can pick up any changes not yet incorporated into the pushed branch.
+	if *pushedRepoURL != *targetRepoURL || *pushedBranch != *targetBranch {
+		targetRef := fmt.Sprintf("%s/%s", gitRemoteName(*targetRepoURL), *targetBranch)
+		if err := git(ctx, &ws.log, "merge", targetRef); err != nil {
+			// Make note of the merge conflict and abort. We'll run all actions and each
+			// one will just fail with the merge conflict error.
+			ws.mergeConflict = true
+			if err := git(ctx, &ws.log, "merge", "--abort"); err != nil {
+				return err
+			}
+		}
+	}
+
 	return nil
 }
 
-func syncExistingRepo(ctx context.Context, path string) error {
-	repo, err := git.PlainOpen(path)
-	if err != nil {
-		return err
+func (ws *workspace) config(ctx context.Context) error {
+	cfg := map[string]string{
+		"user.email":          "ci-runner@buildbuddy.io",
+		"user.name":           "BuildBuddy",
+		"advice.detachedHead": "false",
 	}
-	tree, err := repo.Worktree()
-	if err != nil {
-		return status.WrapErrorf(err, "get worktree")
+	for k, v := range cfg {
+		if err := git(ctx, io.Discard, "config", k, v); err != nil {
+			return status.WrapError(err, "git config")
+		}
 	}
-	if err := tree.Clean(&git.CleanOptions{Dir: true}); err != nil {
-		return status.WrapErrorf(err, "clean")
+	return nil
+}
+
+func (ws *workspace) fetch(ctx context.Context, remoteURL string, branches []string) error {
+	if len(branches) == 0 {
+		return nil
 	}
-	remote, err := repo.Remote(git.DefaultRemoteName)
+	authURL, err := gitutil.AuthRepoURL(remoteURL, os.Getenv(repoUserEnvVarName), os.Getenv(repoTokenEnvVarName))
 	if err != nil {
-		return status.WrapErrorf(err, "get existing remote %q", git.DefaultRemoteName)
-	}
-	if err := fetchOrigin(ctx, remote); err != nil {
 		return err
 	}
-	checkoutOpts := &git.CheckoutOptions{
-		Hash:  gitplumbing.NewHash(*commitSHA),
-		Force: true, // remove local changes
+	remoteName := gitRemoteName(remoteURL)
+	if err := git(ctx, &ws.log, "remote", "add", remoteName, authURL); err != nil && !isRemoteAlreadyExists(err) {
+		return status.WrapError(err, "git remote add")
 	}
-	if err := tree.Checkout(checkoutOpts); err != nil {
-		return status.WrapErrorf(err, "checkout %s", *commitSHA)
+	fetchArgs := append([]string{"fetch", "--force", remoteName}, branches...)
+	if err := git(ctx, &ws.log, fetchArgs...); err != nil {
+		return status.WrapError(err, "git fetch")
 	}
 	return nil
 }
 
+func isRemoteAlreadyExists(err error) bool {
+	return getExitCode(err) == 128
```

Comment:
nit: link to where these codes are defined

Reply:
Looks like these codes aren't special -- git seems to use exit codes 1 and 128 somewhat arbitrarily. Switched to checking the git command output instead, although both approaches seem equally brittle.